technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+
+While we will continue to use and support Helm, it will not be required for installing k8ssandra from a purely technical perspective. Users can install manifests with kubectl apply for example. We can also provide support for kustomize.
+
+## CustomResourceDefinitions
+
+We will utilize a number of CRDs, some of which are developed and maintained by other projects.
+
+There will be a separate controller for each CRD.
+
+The goal of this document is not to be entirely prescriptive for each CRD. Rather, the intent is to give a general idea of how things will look and fit together. 
+
+### K8ssandraCluster
+
+This is the primary object which users will manage. The controller will be responsible for creating other constituent objects.
+
+It will look something like this:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: K8ssandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  cassandra:
+    clusterName: k8ssandra
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
+    enabled: true
+    # optionally set and override defaults for Reaper    
+  stargate: 
+    enabled: true  
+    # optionally set and override defaults for Stargate 
+  monitoring:
+    prometheus:
+      enabled: true
+      # optionally set and override defaults for Prometheus
+    grafana: 
+      enabled: true
+      # optionally set and override defaults for Grafana
+```
+
+Similar to the k8ssandra Helm chart, there is an enabled flag for each component. The Go type for this will be a *bool. In general as it is with the Helm chart, setting enabled: true should be all that is necessary. The controller will take care of configuring the respective objects. 
+
+The user will be able to override defaults as necessary as illustrated in the next example:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: K8ssandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  cassandra:
+    clusterName: k8ssandra
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
+    enabled: true
+    serverConfig:
+      jmxUserSecretName: my-custom-secret
+      cassandraBackend:
+        replication:
+          networkTopology:
+            dc1: 5         
+  stargate: 
+    enabled: true  
+    replicas: 3
+```
+
+### CassandraCluster
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  clusterName: k8ssandra
+  auth:
+    enabled: true
+  datacenters:
+  - name: dc1
+    size: 3
+```
+
+The controller will create a CassandraDatacenter object for each entry in the datacenters array. Similar to what we have done in the Helm charts, you will be able to specify things at cluster level and override them at the DC level. You will also be able to specify properties at the DC level without specifying them at the cluster level.
+
+Multi-cluster is out of scope, but I want to point out that I think this is where we introduce it. Consider the following example:
```

Comment:
With mulit-cluster support now sort of squarely in our sights for K8ssandra, we should definitely consider this aspect as we start making all of these decisions.

Reply:
I have been thinking a lot about multi-cluster over the past few days, particularly around how we do orchestration. I will try to get something written up within the next couple days.